### DISCUSSION TITLE
Problem with using the function of recv/send property

### DISCUSSION THREAD

**Discussion by JPereira1330** [curl/curl#8712] [Q&A] [answered]
I'm developing an app for a embedded system, and inside the library that i have to use is [libcurl][1], the problem is that the embedded system that i am using does not use the write/read function of the ansi C, it requires the use of the property functions, doing so that the [libcurl][1] doesn't function properly in the system.

I searched in the legal documentation and found the tags that could be configure by `curl_easy_setopt` :

 - [CURLOPT_SOCKOPTFUNCTION][2]
 - [CURLOPT_READFUNCTION][3]
 - [CURLOPT_WRITEFUNCTION][4]
 - [CURLOPT_CLOSESOCKETFUNCTION][5]

To my understanding, the cited tags, allow the configuration of the callback to be used after the execution of the respective functions, but not allowing me to alter the form of creating the `socket`, `write`, `read` and `close`.

So, how can I configure the [libcurl][1] to execute property functions?

  [1]: https://curl.se/libcurl/
  [2]: https://curl.se/libcurl/c/CURLOPT_SOCKOPTFUNCTION.html
  [3]: https://curl.se/libcurl/c/CURLOPT_READFUNCTION.html
  [4]: https://curl.se/libcurl/c/CURLOPT_WRITEFUNCTION.html
  [5]: https://curl.se/libcurl/c/CURLOPT_CLOSESOCKETFUNCTION.html

**Comment by jay** [comment]
> the problem is that the embedded system that i am using does not use the write/read function of the ansi C, it requires the use of the property functions, doing so that the [libcurl](https://curl.se/libcurl/) doesn't function properly in the system.

What do you mean by property functions?

**Reply by JPereira1330** [reply]
I believe that I explained it better in the comment below.

**Comment by bagder** [comment]
When is curl using read/write ? Do you mean recv/send ?

**Reply by JPereira1330** [reply]
Yes, sorry I used the wrong terms.

**Comment by JPereira1330** [comment, marked as answer]
The following code works perfectly, when executed in my PC, but when i execute it in the embedded system it doesn't work, resulting in `segmetation fault`.

```c
#include <stdio.h>
#include <stdlib.h>
#include <curl/curl.h>

int main(void) {
    CURL *curl;
    CURLcode res;

    curl = curl_easy_init();
    if (!curl)
        return EXIT_FAILURE;

    curl_easy_setopt(curl, CURLOPT_URL, "https://httpbin.org/ip");
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);

    res = curl_easy_perform(curl);
    if (res != CURLE_OK)
        return EXIT_FAILURE;

    curl_easy_cleanup(curl);
    
    return EXIT_SUCCESS;
}
```

This happens because of certain basic functions of C (Example: `recv()` / `send()`) were completely modified on the hardware, forcing me to use `foo_recv()` / `foo_send()`(created by the hardware's company), because of that I am wondering if it exist a form of using this `foo_recv()` / `foo_send()` instead of  `recv()` / `send()` without having to alter the source code of libcurl.

**Reply by bagder** [reply]
internally, the two macros `sread()` and `swrite()` are used for reading and writing data to and from sockets, "all" you need to do is to make sure they use the correct functions.

Even though it would be even more clever of you to fix your recv/send functions to work like POSIX since that will make it easier for you easier port other source code to your platform as well...
